src/lib.rs

```diff
@@ -103,6 +103,18 @@ pub trait ObjectStore: std::fmt::Display + Send + Sync + Debug + 'static {
     ///
     /// Will return an error if the destination already has an object.
     async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> Result<()>;
+
+    /// Move an object from one path to another in the same object store.
+    ///
+    /// Will return an error if the destination already has an object.
+    async fn rename_if_not_exists(&self, from: &Path, to: &Path) -> Result<()> {
+        match self.copy_if_not_exists(from, to).await {
+            Ok(_) => self.delete(from).await,
+            // TODO do we need to do more rigorous error checking?
+            other => other,
+        }?;
```

Comment:
```suggestion
        self.copy_if_not_exists(from, to).await?;
        self.delete(from).await
```
I don't think you need any additional error checking that isn't already performed by copy_if_not_exists?